KAFKA-14133: Migrate StandbyTaskCreator mock in TaskManagerTest to Mockito

clolov opened the pull request:
This pull requests migrates the StandbyTaskCreator mock in TaskManagerTest from EasyMock to Mockito.
The change is restricted to a single mock to minimize the scope and make it easier for review.

The reasoning as to why we would like to migrate a single mock rather than all mocks in the file at the same time has been discussed in https://github.com/apache/kafka/pull/12607#issuecomment-1500829973

It takes the same approach as in:
https://github.com/apache/kafka/pull/13529
https://github.com/apache/kafka/pull/13621
https://github.com/apache/kafka/pull/13681

**This needs to be rebased on top of https://github.com/apache/kafka/pull/13681 before merging**

clolov committed:
KAFKA-14133: Migrate StandbyTaskCreator mock in TaskManagerTest to Mockito

cadonna commented:
@clolov Could you please do the mentioned rebase?

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -352,16 +356,17 @@ public void shouldRemoveUnusedStandbyTaskFromStateUpdater() {
         final TaskManager taskManager = setUpTaskManager(ProcessingMode.AT_LEAST_ONCE, tasks, true);
-        when(stateUpdater.getTasks()).thenReturn(mkSet(standbyTaskToClose));
+        final Set<Task> standbyTasks = mkSet(standbyTaskToClose);
+        when(stateUpdater.getTasks()).thenReturn(standbyTasks);
```
nit: Why did you introduce variable `standbyTasks`? The variable is not used anywhere else.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1740,13 +1740,14 @@ public void shouldComputeOffsetSumFromCheckpointFileForUninitializedTask() throw
 
         assertThat(taskManager.getTaskOffsetSums(), is(expectedOffsetSums));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
I do not think you need this verification since also with EasyMock the call is not verified (i.e., `verify()` on the mock is never called).

clolov replied:
To be honest, here I followed what was done in PR https://github.com/apache/kafka/pull/13681
For example, the original in the mentioned PR was.
```
...
expect(activeTaskCreator.createTasks(consumer, Collections.emptyMap())).andReturn(emptySet());
...
```
(https://github.com/apache/kafka/pull/13681/files#diff-48bc1476f0437fd711093c7c80ce73eda10be0511705799b4248545c203d521dL294)


and with Mockito we changed it to
```
...
Mockito.verify(activeTaskCreator).createTasks(consumer, Collections.emptyMap());
...
```
(https://github.com/apache/kafka/pull/13681/files#diff-48bc1476f0437fd711093c7c80ce73eda10be0511705799b4248545c203d521dR306)

I followed the same logic for what I thought was a stronger verification, but I am happy to remove it if you think it is unnecessary. I realise the reasoning for doing the above in that PR might have made more sense because there was more than one argument.

Let me know!

clolov replied:
Oh, I see what you mean. You mean that in the EasyMock version there are no calls to
```
verify(standbyTaskCreator)
```
in the these specific tests.
Yeah, sure, I will change it!

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1766,15 +1767,16 @@ public void shouldComputeOffsetSumFromCheckpointFileForClosedTask() throws Excep
 
         assertThat(taskManager.getTaskOffsetSums(), is(expectedOffsetSums));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
Also here, I do not think you need to verify.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1854,6 +1855,8 @@ public void shouldCloseActiveUnassignedSuspendedTasksWhenClosingRevokedTasks() {
         assertThat(taskManager.activeTaskMap(), Matchers.anEmptyMap());
         assertThat(taskManager.standbyTaskMap(), Matchers.anEmptyMap());
+
+        Mockito.verify(standbyTaskCreator, times(2)).createTasks(Collections.emptyMap());
```
Also here no verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1889,6 +1891,8 @@ public void closeClean() {
         );
         assertThat(thrown.getCause().getMessage(), is("KABOOM!"));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1988,6 +1991,8 @@ public void shouldThrowWhenHandlingClosingTasksOnProducerCloseError() {
         assertThat(thrown.getCause(), instanceOf(RuntimeException.class));
         assertThat(thrown.getCause().getMessage(), is("KABOOM!"));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2031,6 +2035,8 @@ public void postCommit(final boolean enforceCheckpoint) {
         verify(stateManager);
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2067,6 +2072,8 @@ public void suspend() {
         verify(stateManager);
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2108,6 +2113,8 @@ public void shouldCommitNonCorruptedTasksOnTaskCorruptedException() {
 
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2144,6 +2149,8 @@ public void shouldNotCommitNonRunningNonCorruptedTasks() {
         assertFalse(nonRunningNonCorruptedTask.commitPrepared);
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2232,6 +2238,8 @@ public void shouldNotAttemptToCommitInHandleCorruptedDuringARebalance() {
         assertThat(uncorruptedActive.state(), is(State.RUNNING));
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2295,6 +2302,8 @@ public void markChangelogAsCorrupted(final Collection<TopicPartition> partitions
         assertThat(uncorruptedActive.state(), is(Task.State.CREATED));
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
See above

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2379,6 +2387,8 @@ public void markChangelogAsCorrupted(final Collection<TopicPartition> partitions
         assertThat(uncorruptedTaskChangelogMarkedAsCorrupted.get(), is(true));
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2433,6 +2442,8 @@ public void markChangelogAsCorrupted(final Collection<TopicPartition> partitions
         assertThat(unrevokedActiveTaskWithCommitNeeded.state(), is(State.CREATED));
         assertThat(unrevokedActiveTaskWithoutCommitNeeded.state(), is(State.RUNNING));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2507,19 +2517,20 @@ public void markChangelogAsCorrupted(final Collection<TopicPartition> partitions
         assertThat(unrevokedActiveTask.state(), is(State.CREATED));
         assertThat(unrevokedActiveTaskWithoutCommitNeeded.state(), is(State.RUNNING));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2556,6 +2567,8 @@ public void shouldAddNonResumedSuspendedTasks() {
 
         verify(activeTaskCreator);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2581,6 +2593,7 @@ public void shouldUpdateInputPartitionsAfterRebalance() {
         assertThat(task00.state(), is(Task.State.RUNNING));
         assertEquals(newPartitionsSet, task00.inputPartitions());
         verify(activeTaskCreator, consumer);
+        Mockito.verify(standbyTaskCreator, times(2)).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2606,6 +2618,7 @@ public void shouldAddNewActiveTasks() {
         assertThat(taskManager.standbyTaskMap(), Matchers.anEmptyMap());
         verify(activeTaskCreator);
         Mockito.verify(changeLogReader).enforceRestoreActive();
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2652,6 +2664,7 @@ public void initializeIfNeeded() {
         assertThat(taskManager.standbyTaskMap(), Matchers.anEmptyMap());
         verify(activeTaskCreator);
         Mockito.verify(changeLogReader).enforceRestoreActive();
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2690,6 +2702,7 @@ public void completeRestoration(final java.util.function.Consumer<Set<TopicParti
         assertThat(taskManager.standbyTaskMap(), Matchers.anEmptyMap());
         verify(activeTaskCreator);
         Mockito.verify(changeLogReader).enforceRestoreActive();
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2700,18 +2713,19 @@ public void shouldSuspendActiveTasksDuringRevocation() {
         taskManager.handleRevocation(taskId00Partitions);
         assertThat(task00.state(), is(Task.State.SUSPENDED));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2880,6 +2893,8 @@ public void shouldNotCommitOnHandleAssignmentIfNoTaskClosed() {
         assertThat(task00.commitNeeded, is(true));
         assertThat(task10.commitPrepared, is(false));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2911,18 +2925,19 @@ public void shouldNotCommitOnHandleAssignmentIfOnlyStandbyTaskClosed() {
 
         assertThat(task00.commitNeeded, is(true));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2932,6 +2947,8 @@ public void shouldNotCommitCreatedTasksOnRevocationOrClosure() {
         taskManager.handleAssignment(emptyMap(), emptyMap());
         assertThat(task00.state(), is(Task.State.CLOSED));
+
+        Mockito.verify(standbyTaskCreator, times(2)).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2956,6 +2972,8 @@ public void suspend() {
 
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3070,6 +3087,8 @@ public void closeDirty() {
         // the active task creator should also get closed (so that it closes the thread producer if applicable)
         verify(activeTaskCreator);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3122,6 +3140,8 @@ public Set<TopicPartition> changelogPartitions() {
         verify(activeTaskCreator);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
     }
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3172,6 +3191,8 @@ public Set<TopicPartition> changelogPartitions() {
         // the active task creator should also get closed (so that it closes the thread producer if applicable)
         verify(activeTaskCreator);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3307,6 +3327,8 @@ public void suspend() {
         // the active task creator should also get closed (so that it closes the thread producer if applicable)
         verify(activeTaskCreator);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3441,6 +3462,8 @@ public void shouldInitializeNewActiveTasks() {
         // verifies that we actually resume the assignment at the end of restoration.
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3716,6 +3738,8 @@ public Map<TopicPartition, OffsetAndMetadata> prepareCommit() {
             assertThrows(RuntimeException.class, () -> taskManager.commitAll());
         assertThat(thrown.getMessage(), equalTo("opsh."));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3780,6 +3803,8 @@ public Map<TopicPartition, Long> purgeableOffsets() {
 
         verify(adminClient);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3818,6 +3842,8 @@ public Map<TopicPartition, Long> purgeableOffsets() {
 
         verify(adminClient);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3958,6 +3983,8 @@ public void shouldProcessActiveTasks() {
         assertThat(taskManager.process(3, time), is(5));
         assertThat(taskManager.process(3, time), is(0));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4046,6 +4072,8 @@ public boolean process(final long wallClockTime) {
 
         assertThrows(TaskMigratedException.class, () -> taskManager.process(1, time));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4075,6 +4102,8 @@ public boolean process(final long wallClockTime) {
         assertThat(exception.taskId().get(), is(taskId00));
         assertThat(exception.getCause().getMessage(), is("oops"));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4088,15 +4117,16 @@ public boolean maybePunctuateStreamTime() {
 
         assertThrows(TaskMigratedException.class, () -> taskManager.punctuate());
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4110,15 +4140,16 @@ public boolean maybePunctuateStreamTime() {
 
         assertThrows(KafkaException.class, () -> taskManager.punctuate());
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4149,6 +4178,8 @@ public boolean maybePunctuateSystemTime() {
         // one for stream and one for system time
         assertThat(taskManager.punctuate(), equalTo(2));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4161,15 +4192,16 @@ public Set<TopicPartition> changelogPartitions() {
         // since the taskManager should _not_ resume the assignment while we're still in RESTORING
         verify(consumer);
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4204,6 +4235,8 @@ public void shouldHaveRemainingPartitionsUncleared() {
             );
         }
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4595,6 +4627,8 @@ public void suspend() {
         assertThat(task00.state(), is(Task.State.SUSPENDED));
         assertThat(task01.state(), is(Task.State.SUSPENDED));
+
+        Mockito.verify(standbyTaskCreator).createTasks(Collections.emptyMap());
```
No verification needed.

cadonna reviewed (commented):
Thanks for the PR, @clolov !

Here my feedback.

clolov commented:
Yup, I will address all comments today, thank you for the review!

clolov committed:
Fixing some stragglers

clolov commented:
Heya @cadonna, I hope I have addressed all of your comments:
* Updated the PR's overview
* Rebased
* Comments on the code itself

Interestingly enough the test `shouldInitializeNewStandbyTasks` fails locally when I run all tests, but succeeds when it is ran in isolation. If the build passes I will blame it on something in my environment, but if it fails I would be glad for another pair of eyes as to what might be causing it.

cadonna commented:
The issue with `shouldInitializeNewStandbyTasks()` seems to be a missed cleanup due to a bug in EasyMock. If I rename the method to `shouldInitialiseNewStandbyTasks()` (notice the s instead of the z) it works for me locally. Probably, the order in which the test methods are executed matters. 
I know, this sounds ridiculous and actually it is ridiculous. :neutral_face:
So rename the method. As soon as we will have migrated the whole test class the issue should be gone anyways.

clolov commented:
Okay, the name has been changed and this has been rebased. I ran checkstyleTest and spotbugsTest and they are passing locally. If everything passes in the automated tests you should be able to merge it at your leisure!

cadonna reviewed (approved):
Thanks, @clolov !

LGTM!

clolov commented:
Good morning! I believe that the failing checks are not connected with this change. Is it okay to merge this PR?

divijvaidya commented:
Failing tests are not related. Merging this code.
```
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testReplicateSourceDefault()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13711/5/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_11_and_Scala_2_13___testReplicateSourceDefault__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testReplicateSourceDefault()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13711/5/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_11_and_Scala_2_13___testReplicateSourceDefault___2/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13711/5/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_17_and_Scala_2_13___testBalancePartitionLeaders__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13711/5/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders__/)
```
